src/compiler/transformers/declarations.ts

```diff
@@ -725,45 +741,197 @@ export function transformDeclarations(context: TransformationContext) {
-            case SyntaxKind.MethodSignature:
-            case SyntaxKind.MethodDeclaration:
-            case SyntaxKind.GetAccessor:
-            case SyntaxKind.CallSignature:
-                typeNode = resolver.createReturnTypeOfSignatureDeclaration(node, enclosingDeclaration, declarationEmitNodeBuilderFlags, symbolTracker);
-                break;
-            default:
-                Debug.assertNever(node);
+
+        function inferReturnTypeOfSignatureSignature<T extends SignatureDeclaration>(node: T) {
+            return resolver.createReturnTypeOfSignatureDeclaration(node, enclosingDeclaration, declarationEmitNodeBuilderFlags, symbolTracker) ?? factory.createKeywordTypeNode(SyntaxKind.AnyKeyword);
         }
+        function inferAccessorType(node: GetAccessorDeclaration | SetAccessorDeclaration) {
+            if (node.kind === SyntaxKind.GetAccessor) {
+                return inferReturnTypeOfSignatureSignature(node);
+            }
+            else {
+                return factory.createKeywordTypeNode(SyntaxKind.AnyKeyword);
+            }
+        }
+        function typeFromDeclaration(node: HasInferredType | ExportAssignment) {
+            const savedErrorNameNode = errorNameNode;
+            const savedErrorFallbackNode = errorFallbackNode;
 
-        errorNameNode = undefined;
-        if (!suppressNewDiagnosticContexts) {
-            getSymbolAccessibilityDiagnostic = oldDiag!;
+            errorNameNode = node.name;
+            errorFallbackNode = node;
+            let oldDiag;
+            if (!suppressNewDiagnosticContexts) {
+                oldDiag = getSymbolAccessibilityDiagnostic;
+                getSymbolAccessibilityDiagnostic = createGetSymbolAccessibilityDiagnosticForNode(node);
+            }
+            let typeNode: TypeNode | undefined;
+            let declaredType: TypeNode | undefined;
+            switch (node.kind) {
+                case SyntaxKind.Parameter:
+                    typeNode = typeFromParameter(node);
+                    break;
+                case SyntaxKind.VariableDeclaration:
+                    typeNode = typeFromVariable(node);
+                    break;
+                case SyntaxKind.PropertyDeclaration:
+                    typeNode = typeFromProperty(node);
+                    break;
+                case SyntaxKind.GetAccessor:
+                    typeNode = typeFromAccessor(node);
+                    break;
+                case SyntaxKind.BindingElement:
+                    typeNode = inferVariableLikeType(node);
+                    break;
+                case SyntaxKind.ConstructSignature:
+                case SyntaxKind.PropertySignature:
+                case SyntaxKind.MethodSignature:
+                case SyntaxKind.CallSignature:
+                    declaredType = node.type;
+                    typeNode = declaredType ? visitTypeNode(declaredType) : factory.createKeywordTypeNode(SyntaxKind.AnyKeyword);

```

Comment:
`visitNode` allows `node` to be undefined. If you add an overload to `visitTypeNode` that allows undefined, this could simply be:

```suggestion
                    typeNode = visitTypeNode(node.type) ?? factory.createKeywordTypeNode(SyntaxKind.AnyKeyword);
```